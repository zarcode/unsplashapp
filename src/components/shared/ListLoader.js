import React from 'react';
import { View, ActivityIndicator, StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  spinnerWrap: {
    height: 50,
    paddingVertical: 15,
  },
});

const ListLoader = ({ loading }: boolean) => (
  <View style={styles.spinnerWrap}>{loading && <ActivityIndicator />}</View>
);

export default ListLoader;
