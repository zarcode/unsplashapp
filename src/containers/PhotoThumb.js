// @flow
import React from 'react';
import { TouchableOpacity, Image } from 'react-native';
import type { PhotoViewModel } from './PhotosList';

type Props = {
  onPress: () => void,
  photo: PhotoViewModel,
  size: number,
}
const PhotoThumb = ({ onPress, photo, size }: Props) => (
  <TouchableOpacity
    style={{
      padding: 1,
    }}
    onPress={onPress}
  >
    <Image
      style={{
        width: size,
        height: size,
      }}
      source={{
        uri: photo.url,
      }}
    />
  </TouchableOpacity>
);

export default PhotoThumb;
