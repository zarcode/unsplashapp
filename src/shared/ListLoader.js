import React from 'react';
import { View, ActivityIndicator, StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  spinnerWrap: {
    paddingVertical: 15,
  },
})

const ListLoader = () => (
  <View style={styles.spinnerWrap}>
    <ActivityIndicator />
  </View>
);

export default ListLoader;
