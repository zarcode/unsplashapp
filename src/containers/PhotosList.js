// @flow
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { TouchableOpacity, FlatList, Image, Dimensions } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

// import PhotoSingleScreen from './PhotoSingleScreen';
import { photosRequested } from '../action/photos';
import {
  getPhotos,
  getFilter,
  getErrorMessage,
  getIsLastPage,
  getLastLoadedPage,
  getLoadingState,
} from '../reducers/photos';
import type { Photo, PhotosFilter } from '../api/types';
import ListLoader from '../shared/ListLoader';
// import {List} from "immutable";
// onPress={() => this.props.navigation.dispatch({ type: 'PhotoSingleScreen' })}

type Props = {
  navigation: any,
  photos: Array<Photo>,
  filter: PhotosFilter,
  isLastPage: boolean,
  lastLoadedPage: number,
  loadingState: 'refreshing' | 'loading' | 'idle',
  actions: {
    photosRequested: (filter: PhotosFilter, refresh: boolean) => void
  }
};

type State = {
  numColumns: number,
  imageDim: number,
};

class PhotosList extends Component<Props, State> {
  constructor(props) {
    super(props);

    this.state = {
      numColumns: 3,
      imageDim: Math.floor(Dimensions.get('window').width / 3),
    };
  }

  state: State;

  componentDidMount() {
    this.onStart();
  }

  onPress = () => {}

  onLayout = () => {
    const { width, height } = Dimensions.get('window');
    let columns = 3;
    if (width > 1920) {
      columns = 5;
    }
    if (width > height) {
      columns = 5;
      if (width > 1920) {
        columns = 10;
      }
    }
    this.setState({ numColumns: columns, imageDim: Math.floor(width / columns) });
  };

  // onRefresh = () => {
  //   this.loadIfRequired(true);
  //   this.setState({
  //     refreshing: true,
  //   });
  // }
  //
  // onLoadMore = () => {
  //   if (!this.props.loadingState && !this.state.refreshing) {
  //     this.loadIfRequired(false);
  //   }
  // }
  //
  // loadIfRequired = (refresh: boolean): void => {
  //   this.props.actions.photosRequested(this.props.filter, refresh);
  // }

  onStart = () => {
    const isLoaded = this.props.lastLoadedPage > 0;
    const isIdle = this.props.loadingState === 'idle';
    // const isRestored = this.props.isRestored;
    const isRestored = false;
    if (isIdle && (!isLoaded || isRestored)) {
      this.props.actions.photosRequested(this.props.filter, true);
    }
  };

  refresh = () => {
    this.props.actions.photosRequested(this.props.filter, true);
  };

  loadMore = () => {
    const isLoaded = this.props.lastLoadedPage > 0;
    const isIdle = this.props.loadingState === 'idle';
    const { isLastPage } = this.props;

    if (isLoaded && isIdle && !isLastPage) {
      this.props.actions.photosRequested(this.props.filter, false);
    }
  };

  photoViewModel = item => ({
    id: item.id,
    url: item.urls.small,
  });

  keyExtractor = item => item.id;

  renderItem = ({ item }) => {
    const photo = this.photoViewModel(item);
    return (
      <TouchableOpacity onPress={this.onPress}>
        <Image
          style={{
            width: this.state.imageDim,
            height: this.state.imageDim,
          }}
          source={{
            uri: photo.url,
          }}
        />
      </TouchableOpacity>
    );
  };

  renderFooter = () => {
    if (this.props.loadingState === 'loading') {
      return <ListLoader />;
    }
    return null;
  }

  render() {
    const loading = this.props.loadingState === 'loading';
    const refreshing = this.props.loadingState === 'refreshing';

    return (
      <FlatList
        loading={loading}
        onLayout={this.onLayout}
        key={this.state.numColumns}
        numColumns={this.state.numColumns}
        data={this.props.photos}
        keyExtractor={this.keyExtractor}
        renderItem={this.renderItem}
        ListFooterComponent={this.renderFooter}
        refreshing={refreshing}
        onRefresh={this.refresh}
        onEndReachedThreshold={0.5}
        onEndReached={this.loadMore}
      />
    );
  }
}

PhotosList.propTypes = {
  actions: PropTypes.shape({
    photosRequested: PropTypes.func.isRequired,
  }).isRequired,
};

const mapStateToProps = (state) => {
  const filter = getFilter(state);
  return {
    photos: getPhotos(state),
    filter,
    isLastPage: getIsLastPage(state),
    lastLoadedPage: getLastLoadedPage(state)(filter),
    loadingState: getLoadingState(state),
    getErrorMessage: getErrorMessage(state),
  };
}

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ photosRequested }, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(PhotosList);
