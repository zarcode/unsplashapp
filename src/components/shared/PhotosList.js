// @flow
import React, { Component } from 'react';
import {
  FlatList,
  Dimensions,
  StyleSheet,
  Image,
  Alert,
  PixelRatio,
} from 'react-native';
import propPath from 'crocks/Maybe/propPath';

import type { Photo, PhotosFilter, PhotoID } from '../../api/types';
import ListLoader from '../shared/ListLoader';
import PhotoThumb from './PhotoThumb';

const noImages = require('../../assets/icons/no-images.png');

type Props = {
  photos: Array<Photo>,
  filter: PhotosFilter,
  isLastPage: boolean,
  lastLoadedPage: number,
  loadingState: 'refreshing' | 'loading' | 'idle',
  getErrorMessage: string,
  navigation: any,
  actions: {
    // toSinglePhoto: (photo: Photo) => void,
    photosRequested: (filter: PhotosFilter, refresh: boolean) => void,
    resetPhotos?: () => void,
  },
};

type State = {
  numColumns: number,
  imageDim: number,
  imagePixels: number,
  alertVisible: boolean,
};

export type PhotoViewModel = {
  id: PhotoID,
  url: string,
};

const styles = StyleSheet.create({
  listContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default class PhotosListComponent extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    const imageDim = Dimensions.get('window').width / 3;
    this.state = {
      numColumns: 3,
      imageDim: Math.floor(imageDim),
      imagePixels: PixelRatio.getPixelSizeForLayoutSize(imageDim),
      alertVisible: false,
    };
  }

  state: State;

  componentDidMount() {
    this.onStart();
  }
  componentWillReceiveProps(nextProps: Props) {
    if (
      this.props.getErrorMessage === null &&
      nextProps.getErrorMessage &&
      !this.state.alertVisible
    ) {
      this.setState({ alertVisible: true });
      if (nextProps.getErrorMessage === 'limit') {
        Alert.alert(
          'Usage limit exceeded',
          'This is just a code showcase app. Try it again in one hour.',
          [{ text: 'OK', onPress: this.onCloseAlert }],
        );
      } else {
        Alert.alert(
          'Sorry, something went wrong',
          'Please, try refreshing the list',
          [{ text: 'OK', onPress: this.onCloseAlert }],
        );
      }
    }
  }
  componentWillUnmount() {
    if (this.props.actions.resetPhotos) this.props.actions.resetPhotos();
  }
  onCloseAlert = () => {
    this.setState({ alertVisible: false });
  };
  onLayout = () => {
    const { width, height } = Dimensions.get('window');
    let columns = 3;
    if (width > 1920) {
      columns = 5;
    }
    if (width > height) {
      columns = 5;
      if (width > 1920) {
        columns = 10;
      }
    }
    const imageDim = Math.floor(width / columns);
    this.setState({
      numColumns: columns,
      imageDim,
      imagePixels: PixelRatio.getPixelSizeForLayoutSize(imageDim),
    });
  };
  onStart = () => {
    const isLoaded = this.props.lastLoadedPage > 0;
    const isIdle = this.props.loadingState === 'idle';
    // const isRestored = this.props.isRestored;
    const isRestored = false;
    if (isIdle && (!isLoaded || isRestored)) {
      this.props.actions.photosRequested(this.props.filter, true);
    }
  };
  itemLayout = (_: *, index: number) => ({
    length: this.state.imageDim,
    offset: this.state.imageDim * index,
    index,
  });
  refresh = () => {
    this.props.actions.photosRequested(this.props.filter, true);
  };

  loadMore = () => {
    const isLoaded = this.props.lastLoadedPage > 0;
    const isIdle = this.props.loadingState === 'idle';
    const { isLastPage } = this.props;

    if (isLoaded && isIdle && !isLastPage) {
      this.props.actions.photosRequested(this.props.filter, false);
    }
  };

  photoViewModel = (item: Photo): PhotoViewModel => {
    const { imagePixels } = this.state;
    return {
      id: propPath(['id'], item).option(''),
      url: propPath(
        ['urls', imagePixels > 200 ? 'small' : 'thumb'],
        item,
      ).option(null),
    };
  };

  keyExtractor = (item: Photo) => item.id;

  navigateToSingle = (item: Photo) => () => {
    this.props.navigation.navigate('PhotoSingleScreen', {
      photo: item,
    });
  };

  renderItem = ({ item }: Photo) => {
    const photo: PhotoViewModel = this.photoViewModel(item);
    return (
      <PhotoThumb
        onPress={this.navigateToSingle(item)}
        photo={photo}
        size={this.state.imageDim - 2}
      />
    );
  };

  renderFooter = () => {
    if (this.props.loadingState === 'loading') {
      return <ListLoader />;
    }
    return null;
  };

  renderEmpty = () => {
    if (this.props.loadingState !== 'idle') {
      return null;
    }

    return (
      <Image
        style={{
          width: 150,
          height: 150,
          opacity: 0.1,
        }}
        resizeMode="cover"
        source={noImages}
      />
    );
  };

  render() {
    const loading = this.props.loadingState === 'loading';
    const refreshing = this.props.loadingState === 'refreshing';
    return (
      <FlatList
        contentContainerStyle={
          this.props.photos.length === 0 && styles.listContainer
        }
        getItemLayout={this.itemLayout}
        loading={loading}
        onLayout={this.onLayout}
        key={this.state.numColumns}
        numColumns={this.state.numColumns}
        data={this.props.photos}
        keyExtractor={this.keyExtractor}
        renderItem={this.renderItem}
        ListEmptyComponent={this.renderEmpty}
        ListFooterComponent={this.renderFooter}
        refreshing={refreshing}
        onRefresh={this.refresh}
        onEndReachedThreshold={0.2}
        onEndReached={this.loadMore}
      />
    );
  }
}
