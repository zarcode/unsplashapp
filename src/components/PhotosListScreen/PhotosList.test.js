import React from 'react';
import { Image } from 'react-native';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { PhotosListComponent as PhotosList } from './PhotosList';
// import type {Photo} from '../api/types';
// import {AppWithNavigationState} from '../Root';
import ListLoader from '../shared/ListLoader';
import PhotoThumb from './PhotoThumb';

jest.mock('Dimensions');

describe('<PhotosList />', () => {
  const props = {
    filter: 'latest',
    isLastPage: false,
    loadingState: 'idle',
    lastLoadedPage: 1,
    actions: {
      photosRequested: jest.fn(),
    },
  };
  const photo = {
    id: '1',
    urls: {
      small: 'url',
    },
  };
  const wrapper = shallow(<PhotosList {...props} />);
  it('setProps calls componentWillReceiveProps', () => {
    jest.spyOn(PhotosList.prototype, 'componentWillReceiveProps');
    wrapper.setProps({ getErrorMessage: 'Error' });
    expect(PhotosList.prototype.componentWillReceiveProps.mock.calls.length).toBe(1);
  });
  it('componentDidMount calls photosRequested', () => {
    const props1 = {
      loadingState: 'idle',
      lastLoadedPage: 0,
      filter: 'latest',
      actions: {
        photosRequested: jest.fn(),
      },
    };
    jest.spyOn(PhotosList.prototype, 'componentDidMount');
    shallow(<PhotosList {...props1} />);
    // wrapper.instance().onStart();
    expect(PhotosList.prototype.componentDidMount).toBeCalled();
    expect(props1.actions.photosRequested).toBeCalledWith('latest', true);
  });
  it('refresh calls photosRequested', () => {
    wrapper.instance().refresh();
    expect(props.actions.photosRequested).toBeCalledWith('latest', true);
  });
  it('loadMore calls photosRequested', () => {
    wrapper.instance().loadMore();
    expect(props.actions.photosRequested).toBeCalledWith('latest', false);
  });
  it('creates photo model', () => {
    expect(wrapper.instance().photoViewModel(photo)).toEqual({
      id: '1',
      url: 'url',
    });
    const photo1 = {
      id: '1',
    };
    expect(wrapper.instance().photoViewModel(photo1)).toEqual({
      id: '1',
      url: null,
    });
  });
  it('render item', () => {
    const toSinglePhoto = jest.fn();
    const showProps = {
      actions: {
        toSinglePhoto,
      },
    };

    const showStateWrapper = shallow(<PhotosList {...showProps} />);
    showStateWrapper.setState({ imageDim: 500 });
    expect(showStateWrapper.instance().itemLayout({}, 2)).toEqual({
      length: 500,
      offset: 500 * 2,
      index: 2,
    });
    expect(showStateWrapper.instance().renderItem({ photo })).toEqual(<PhotoThumb
      onPress={toSinglePhoto}
      photo={{
          id: 0,
          url: null,
        }}
      size={500 - 2}
    />);
  });
  it('render loader', () => {
    const loadingProps = {
      loadingState: 'loading',
    };

    expect(wrapper.instance().renderFooter()).toEqual(null);
    const loadingStateWrapper = shallow(<PhotosList {...loadingProps} />);
    expect(loadingStateWrapper.instance().renderFooter()).toEqual(<ListLoader />);
    expect(loadingStateWrapper.instance().renderEmpty()).toEqual(null);
    expect(toJson(loadingStateWrapper)).toMatchSnapshot();
  });
  it('render empty', () => {
    const emptyProps = {
      loadingState: 'idle',
      photos: [],
      actions: {
        photosRequested: jest.fn(),
      },
    };
    const emptyStateWrapper = shallow(<PhotosList {...emptyProps} />);
    expect(emptyStateWrapper.instance().renderEmpty()).toEqual(<Image
      resizeMode="cover"
      source={{
          testUri: '../../../src/assets/icons/no-images.png',
        }}
      style={{
          height: 150,
          tintColor: 'rgba(255,255,255,0.4)',
          width: 150,
        }}
    />);
  });
  it('layout is correct', () => {
    wrapper.instance().onLayout();
    expect(wrapper.instance().state).toEqual({
      imageDim: 80,
      numColumns: 5,
    });
  });
});
