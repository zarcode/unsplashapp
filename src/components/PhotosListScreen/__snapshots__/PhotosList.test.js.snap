// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotosList /> render loader 1`] = `
<FlatList
  ListEmptyComponent={[Function]}
  ListFooterComponent={[Function]}
  contentContainerStyle={false}
  disableVirtualization={false}
  getItemLayout={[Function]}
  horizontal={false}
  initialNumToRender={10}
  key="3"
  keyExtractor={[Function]}
  loading={true}
  maxToRenderPerBatch={10}
  numColumns={3}
  onEndReached={[Function]}
  onEndReachedThreshold={0.5}
  onLayout={[Function]}
  onRefresh={[Function]}
  refreshing={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
